src/passes/CodePushing.cpp

```diff
@@ -237,6 +244,178 @@ class Pusher {
+  //
+  // This does not move the push point, so it does not have a return value,
+  // unlike optimizeSegment.
+  void optimizeIntoIf(Index firstPushable, Index pushPoint) {
+    assert(firstPushable != Index(-1) && pushPoint != Index(-1) &&
+           firstPushable < pushPoint);
+
+    auto* iff = list[pushPoint]->dynCast<If>();
+    if (!iff) {
+      return;
+    }
+
+    // Everything that matters if you want to be pushed past the pushPoint. This
+    // begins with the if condition's effects, as we must always push past
+    // those. Later, we will add to this when we need to.
+    EffectAnalyzer cumulativeEffects(passOptions, module, iff->condition);
+
+    // See optimizeSegment for why we can ignore control flow transfers here.
+    cumulativeEffects.ignoreControlFlowTransfers();
+
+    // Find the effects of the arms, which will affect what can be pushed.
+    EffectAnalyzer ifTrueEffects(passOptions, module, iff->ifTrue);
+    EffectAnalyzer ifFalseEffects(passOptions, module);
+    if (iff->ifFalse) {
+      ifFalseEffects.walk(iff->ifFalse);
+    }
+
+    // We need to know which locals are used after the if, as that can determine
+    // if we can push or not.
+    EffectAnalyzer postIfEffects(passOptions, module);
+    for (Index i = pushPoint + 1; i < list.size(); i++) {
+      postIfEffects.walk(list[pushPoint]);
+    }
+
+    // Start at the instruction right before the push point, and go back from
+    // there:
+    //
+    //    x = op();
+    //    y = op();
+    //    if (..) {
+    //      ..
+    //    }
+    //
+    // Here we will try to push y first, and then x. Note that if we push y
+    // then we can immediately try to push x after it, as it will remain in
+    // order with x if we do. If we do *not* push y we can still try to push x
+    // but we must move it past y, which means we need to check for interference
+    // between them (which we do by adding y's effects to cumulativeEffects).
+    //
+    // Decrement at the top of the loop for simplicity, so start with i at one
+    // past the first thing we can push (which is right before the push point).
+    Index i = pushPoint;
+    while (1) {
+      if (i == firstPushable) {
+        // We just finished processing the first thing that could be pushed;
+        // stop.
+        break;
+      }
+      assert(i > 0);
+      i--;
+      auto* pushable = isPushable(list[i]);
+      if (!pushable) {
+        // Something that is staying where it is, so anything we push later must
+        // move past it. Note the effects and continue.
+        cumulativeEffects.walk(list[i]);
+        continue;
+      }
+
+      auto index = pushable->index;
+
+      auto iter = pushableEffects.find(pushable);
+      if (iter == pushableEffects.end()) {
+        iter = pushableEffects
+                 .emplace(std::piecewise_construct,
+                          std::forward_as_tuple(pushable),
+                          std::forward_as_tuple(passOptions, module, pushable))
```

Comment:
Is all this verbosity worth it? We could use `emplace({pushable, EffectAnalyzer(passOptions, module, pushable)})` and it would be much more readable. This could also be pulled out into a helper function.

Reply:
Hmm, I copied this from the other function. Trying it your way, I get
```cpp
/home/azakai/Dev/2-binaryen/src/passes/CodePushing.cpp:331:39: error: no matching function for call to ‘std::unordered_map<wasm::LocalSet*, wasm::EffectAnalyzer>::emplace(<brace-enclosed initializer list>)’
  331 |         iter = pushableEffects.emplace({pushable, EffectAnalyzer(passOptions, module, pushable)}).first;
      |                ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I guess that's why the other function does it... That's more C++ than I understand, to be honest, not sure why it's unhappy. But I can do a helper function, that sounds good.

Reply:
Ugh. I wonder if using `std::make_pair` instead of `{}` would work.

Reply:
It also looks like there is a method [`try_emplace`](https://en.cppreference.com/w/cpp/container/unordered_map/try_emplace) that builds in all this `piecewise_construct` and `forward_as_tuple` junk so you don't have to think about it.

Reply:
Nice, looks like it builds with `try_emplace`...